File: code/game/Geology.cpp
#include "Geology.hpp"
#include "BaalExceptions.hpp"
#include "WorldTile.hpp"

#include <iomanip>

using namespace baal;

///////////////////////////////////////////////////////////////////////////////
Geology::Geology(float base_tension_buildup,
                 float base_magma_buildup,
                 float plate_movement)
///////////////////////////////////////////////////////////////////////////////
  : m_tension(0.0),
    m_magma(0.0),
    m_plate_movement(plate_movement),
    m_tension_buildup(base_tension_buildup * plate_movement),
    m_magma_buildup(base_magma_buildup * plate_movement)
{
  Require(plate_movement >= 0.0,       "Broken precondition");
  Require(base_tension_buildup >= 0.0, "Broken precondition");
  Require(base_magma_buildup >= 0.0,   "Broken precondition");
}

///////////////////////////////////////////////////////////////////////////////
void Geology::cycle_turn()
///////////////////////////////////////////////////////////////////////////////
{
  // Tension/magma build up more slowly as they reach 100%
  m_tension += (1 - m_tension) * m_tension_buildup;
  m_magma   += (1 - m_magma)   * m_magma_buildup;

  Require(m_tension < 1.0, "Invariant violated: " << m_tension);
  Require(m_magma   < 1.0, "Invariant violated: " << m_magma);
}

///////////////////////////////////////////////////////////////////////////////
void Geology::draw_text(std::ostream& out) const
///////////////////////////////////////////////////////////////////////////////
{
  const char* invalid_color = "";
  const char* color = invalid_color;
  std::string invalid_symbol = "????";
  std::string symbol = invalid_symbol;
  float property;
  std::ostringstream oss;

  switch (s_draw_mode) {
  case GEOLOGY:
    color = this->color();
    symbol = this->symbol();
    break;
  case TENSION:
  case MAGMA:
    property = s_draw_mode == TENSION ? m_tension : m_magma;
    if (property < .333) {
      color = GREEN;
    }
    else if (property < .666) {
      color = YELLOW;
    }
    else {
      color = RED;
    }
    oss << std::setprecision(3) << std::fixed << property;
    symbol = oss.str();
    break;
  default:
    Require(false, "Should not draw geology in mode: " << s_draw_mode);
  }

  Require(color  != invalid_color , "Color was not set");
  Require(symbol != invalid_symbol, "Symbol was not set");
  Require(symbol.size() <= WorldTile::TILE_TEXT_WIDTH,
          "Symbol '" << symbol << "' is too long");

  out << BOLD_COLOR << color    // set color and bold text
      << std::setw(WorldTile::TILE_TEXT_WIDTH) << symbol // print symbol
      << CLEAR_ALL;             // clear color and boldness
}

///////////////////////////////////////////////////////////////////////////////
bool Geology::is_geological(DrawMode mode)
///////////////////////////////////////////////////////////////////////////////
{
  return mode == GEOLOGY ||
         mode == TENSION ||
         mode == MAGMA;
}

///////////////////////////////////////////////////////////////////////////////
xmlNodePtr Geology::to_xml()///////////////////////////////////////////////////////////////////////////////
{
  xmlNodePtr Geology_node = xmlNewNode(NULL, BAD_CAST "Geology");

  std::ostringstream m_plate_movement_oss;
  m_plate_movement_oss << m_plate_movement;
  xmlNewChild(Geology_node, NULL, BAD_CAST "m_plate_movement", BAD_CAST m_plate_movement_oss.str().c_str());
  xmlNewChild(Geology_node, NULL, BAD_CAST "type", BAD_CAST this->geology_type());

  return Geology_node;
}

File: code/game/geology.py
#! /usr/bin/env python

"""
This file contains the classes having to do with Geology. The
classes here are just data-holders for the most part. As usual, we define
multiple classes here to avoid having a large number of header files for
very small classes.
"""

import unittest

from drawable import Drawable, DrawMode, curr_draw_mode, _set_draw_mode
from baal_common import prequire, ProgramError, cprint, \
                        GREEN, YELLOW, BLUE, RED, WHITE
from world_tile import WorldTile

#
# Free-function API
#

###############################################################################
def is_geological(draw_mode):
###############################################################################
    return draw_mode in [DrawMode.GEOLOGY, DrawMode.TENSION, DrawMode.MAGMA]

###############################################################################
class Geology(Drawable):
###############################################################################
    """
    Contains geology (plate-tectonic) data. This is an abstract base class for
    all Geology classes.

    Plate boundaries always fall between tiles. Boundaries can be:
    Divergent
    Convergent(ocean/land)
    Convergent(ocean/ocean)
    Convergent(land/land)
    Transform

    Every tile builds up plate tension and magma based on geology.
    """

    #
    # Public API
    #

    ###########################################################################
    def __init__(self, plate_movement):
    ###########################################################################
        self.__tension = 0.0
        self.__magma   = 0.0
        self.__plate_movement = plate_movement
        self.__tension_buildup = self._base_tension_buildup() * plate_movement
        self.__magma_buildup = self._base_magma_buildup() * plate_movement

        prequire(plate_movement >= 0.0, "Broken precondition")
        prequire(self._base_tension_buildup() >= 0.0, "Broken precondition")
        prequire(self._base_magma_buildup() >= 0.0, "Broken precondition")

    #
    # Getters / Queries
    #

    def tension(self): return self.__tension

    def magma(self): return self.__magma

    def to_xml(self): return self.__to_xml_impl()

    #
    # Drawing API
    #

    def draw_text(self): return self.__draw_text_impl()

    def draw_graphics(self): return self.__draw_graphics_impl()

    #
    # Modification API
    #

    def cycle_turn(self):
        """
        Notify this object that the turn has cycled
        """
        return self.__cycle_turn_impl()

    #
    # Protected Abstract API
    #

    @classmethod
    def _color(cls): prequire(False, "Must override")

    @classmethod
    def _symbol(cls): prequire(False, "Must override")

    @classmethod
    def _base_magma_buildup(cls): prequire(False, "Must override")

    @classmethod
    def _base_tension_buildup(cls): prequire(False, "Must override")

    #
    # Implementation
    #

    ###########################################################################
    def __to_xml_impl(self):
    ###########################################################################
        pass # Aaron TODO

    ###########################################################################
    def __draw_text_impl(self):
    ###########################################################################
        draw_mode = curr_draw_mode()
        expected_length = WorldTile.TILE_TEXT_WIDTH

        # Figure out what string to print and what color it should be
        if (draw_mode == DrawMode.GEOLOGY):
            color = self._color()
            to_draw = self._symbol().center(expected_length)
        elif (draw_mode in [DrawMode.TENSION, DrawMode.MAGMA]):
            value = self.__tension if draw_mode == DrawMode.TENSION \
                    else self.__magma
            to_draw = ("%.3f" % value).center(expected_length)

            if (value < .333):
                color = GREEN
            elif (value < .666):
                color = YELLOW
            else:
                color = RED
        else:
            prequire(False, "Should not draw geology in mode: ", draw_mode)

        prequire(len(to_draw) == WorldTile.TILE_TEXT_WIDTH,
                 "Symbol '", to_draw, "' is wrong length")

        cprint(color, to_draw)

    ###########################################################################
    def __draw_graphics_impl(self):
    ###########################################################################
        # TODO
        pass

    ###########################################################################
    def __cycle_turn_impl(self):
    ###########################################################################
        # Tension/magma build up more slowly as they reach 100%
        self.__tension += (1 - self.__tension) * self.__tension_buildup
        self.__magma   += (1 - self.__magma)   * self.__magma_buildup

        prequire(self.__tension < 1.0, "Invariant violated: ", self.__tension)
        prequire(self.__magma   < 1.0, "Invariant violated: ", self.__magma)

#
# Geology classes
#

###############################################################################
class Divergent(Geology):
###############################################################################
    """
    Diverging plate boundary.
    """

    _MAGMA_BUILDUP   = 0.001
    _TENSION_BUILDUP = 0.000
    _COLOR           = BLUE
    _SYMBOL          = "<-->"

    def __init__(self, plate_movement):
        super(self.__class__, self).__init__(plate_movement)

    @classmethod
    def _color(cls): return cls._COLOR

    @classmethod
    def _symbol(cls): return cls._SYMBOL

    @classmethod
    def _base_magma_buildup(cls): return cls._MAGMA_BUILDUP

    @classmethod
    def _base_tension_buildup(cls): return cls._TENSION_BUILDUP

###############################################################################
class Subducting(Geology):
###############################################################################
    """
    There is no difference between convergent land-ocean and convergent
    ocean-ocean. Both are subducting with one plate going under the other.
    """

    _MAGMA_BUILDUP   = 0.002
    _TENSION_BUILDUP = 0.002
    _COLOR           = RED
    _SYMBOL          = "-v<-"

    def __init__(self, plate_movement):
        super(self.__class__, self).__init__(plate_movement)

    @classmethod
    def _color(cls): return cls._COLOR

    @classmethod
    def _symbol(cls): return cls._SYMBOL

    @classmethod
    def _base_magma_buildup(cls): return cls._MAGMA_BUILDUP

    @classmethod
    def _base_tension_buildup(cls): return cls._TENSION_BUILDUP

###############################################################################
class Orogenic(Geology):
###############################################################################
    """
    Coverging land-land => Orogenic boundary. Neither plate goes under.
    """

    _MAGMA_BUILDUP   = 0.000
    _TENSION_BUILDUP = 0.002
    _COLOR           = GREEN
    _SYMBOL          = "-><-"

    def __init__(self, plate_movement):
        super(self.__class__, self).__init__(plate_movement)

    @classmethod
    def _color(cls): return cls._COLOR

    @classmethod
    def _symbol(cls): return cls._SYMBOL

    @classmethod
    def _base_magma_buildup(cls): return cls._MAGMA_BUILDUP

    @classmethod
    def _base_tension_buildup(cls): return cls._TENSION_BUILDUP

###############################################################################
class Transform(Geology):
###############################################################################
    """
    Transform - Two plates sliding past each other.
    """

    _MAGMA_BUILDUP   = 0.000
    _TENSION_BUILDUP = 0.003
    _COLOR           = YELLOW
    _SYMBOL          = "vv^^"

    def __init__(self, plate_movement):
        super(self.__class__, self).__init__(plate_movement)

    @classmethod
    def _color(cls): return cls._COLOR

    @classmethod
    def _symbol(cls): return cls._SYMBOL

    @classmethod
    def _base_magma_buildup(cls): return cls._MAGMA_BUILDUP

    @classmethod
    def _base_tension_buildup(cls): return cls._TENSION_BUILDUP

###############################################################################
class Inactive(Geology):
###############################################################################
    """
    Inactive - No nearby plate boundary
    """

    _MAGMA_BUILDUP   = 0.000
    _TENSION_BUILDUP = 0.000
    _COLOR           = WHITE
    _SYMBOL          = ""

    def __init__(self):
        super(self.__class__, self).__init__(0.0)

    @classmethod
    def _color(cls): return cls._COLOR

    @classmethod
    def _symbol(cls): return cls._SYMBOL

    @classmethod
    def _base_magma_buildup(cls): return cls._MAGMA_BUILDUP

    @classmethod
    def _base_tension_buildup(cls): return cls._TENSION_BUILDUP

#
# Tests
#

###############################################################################
class TestGeology(unittest.TestCase):
###############################################################################

    ###########################################################################
    def test_geology(self):
    ###########################################################################
        plate_movement = 3

        # Test that we cannot create instances of Geology
        self.assertRaises(ProgramError, Geology, plate_movement)

        # Check that classmethod structure works
        subd = Subducting(plate_movement)
        self.assertEqual(subd._color(), Subducting._COLOR)

        # Check that drawing works
        _set_draw_mode(DrawMode.CIV)
        self.assertRaises(ProgramError, subd.draw_text)
        self.assertTrue(is_geological(DrawMode.MAGMA))
        _set_draw_mode(DrawMode.MAGMA)
        subd.draw_text()
        _set_draw_mode(DrawMode.TENSION)
        subd.draw_text()

        # Check turn cycling
        subd.cycle_turn()
        subd.cycle_turn()
        tension1 = subd.tension()
        magma1 = subd.magma()
        subd.cycle_turn()
        subd.cycle_turn()
        self.assertLess(tension1, subd.tension())
        self.assertLess(magma1, subd.magma())

        for i in xrange(10000):
            subd.cycle_turn()
        self.assertLess(subd.tension(), 1.0)
        self.assertLess(subd.magma(), 1.0)

        # Check transform
        trans = Transform(plate_movement)
        magma_orig = trans.magma()
        trans.cycle_turn()
        trans.cycle_turn()
        tension1 = trans.tension()
        self.assertEqual(magma_orig, trans.magma())
        trans.cycle_turn()
        trans.cycle_turn()
        self.assertLess(tension1, trans.tension())
        self.assertEqual(magma_orig, trans.magma())

        # Check inactive
        inactive = Inactive()
        magma_orig = inactive.magma()
        tension_orig = inactive.tension()
        for i in xrange(1000):
            inactive.cycle_turn()
        self.assertEqual(magma_orig, inactive.magma())
        self.assertEqual(tension_orig, inactive.tension())

if (__name__ == "__main__"):
    unittest.main()
